### DISCUSSION TITLE
ExternalDomain vs zitadel react tutorial

### DISCUSSION THREAD

**Discussion by bernardro** [zitadel/zitadel#6772] [Q&A]
Hi, based on the zitadel tutorial [here](https://www.youtube.com/watch?v=5THbQljoPKg) I have:
- zitadel set up at https://zitadel.mysvr.com (local docker behind traefik with ExternalDomain=zitadel.mysvr.com)
- the [zitadel react sample app](https://github.com/zitadel/react-user-authentication) duplicated at https://zitareact1.mysvr.com and https://zitareact2.mysvr.com

how can I get the standard signup flow to work for the react apps? (while keeping ExternalDomain=zitadel.mysvr.com)

**Comment by bernardro** [comment]
after setting everything up per the tutorial including the redirect urls and `authConfig.js` in the react apps themselves, when I go to https://zitareact1.mysvr.com/ it automatically gets redirected to https://zitareact1.mysvr.com/ui/login - which then gives error below
> ID=QUERY-1kIjX Message=Instance not found

I am using the zitadel+traefik sample code [here](https://zitadel.com/docs/self-hosting/deploy/loadbalancing-example):
zitadel version v2.39.1
cockroachDb v23.1.11

**Comment by hifabienne** [comment]
He @bernardro 
Not sure how your react app config looks like.
In the authConfig you have to configured needs the right url.
https://github.com/zitadel/react-user-authentication/blob/main/src/authConfig.js
Authority should be https://zitadel.mysvr.com/
Also the userinfo should point to that url

The redirect url should be something in your app so https://zitareact1.mysvr.com/...

**Reply by bernardro** [reply]
yep, that's exactly how I have it
```
const authConfig = {
    authority: 'https://zitadel.mysvr.com', //Replace with your issuer URL
    client_id: 'REDACTED@demologin', //Replace with your client id
    redirect_uri: 'https://zitareact1.mysvr.com/callback',
    response_type: 'code',
    scope: 'openid profile email',
    post_logout_redirect_uri: 'https://zitareact1.mysvr.com/',
    userinfo_endpoint: 'https://zitadel.mysvr.com/oidc/v1/userinfo', //Replace with your user-info endpoint
    response_mode: 'query',
    code_challenge_method: 'S256',
  };
```

**Reply by bernardro** [reply]
zitadel is clearly seeing the attempt to browse zitareact1.mysvr.com - and is thwarting it or failing to recognize that the domain does exist in zitadel, I have even added `zitareact1.mysvr.com` to the list of "verified domains" under Organization > settings

**Reply by bernardro** [reply]
zitadel-config.yaml
```
...
ExternalSecure: true
ExternalDomain: zitadel.mysvr.com
ExternalPort: 443
...
```

**Comment by hifabienne** [comment]
@eliobischof Maybe you can help here?

**Reply by bernardro** [reply]
latest effort: in my docker-compose.yaml I did this
```
...
    environment:
      - ZITADEL_HTTP2HOSTHEADER="authoritytwo"
      - ZITADEL_HTTP1HOSTHEADER="authorityone"
...
        traefik.http.middlewares.zitadel.headers.customrequestheaders.authoritytwo: "zitadel.mysvr.local"
        traefik.http.middlewares.zitadel.headers.customrequestheaders.authorityone: "zitadel.mysvr.local"
...
```
then when I tried to browse my `zitadel.mysvr.local`, I got 
>host header "authorityone" not found

also when I try to open `zitareact1.mysvr.local` I get the usual 
>Message=Instance not found

but when I use the IP address (not reverse proxied) e.g ` http://172.18.0.3:7000` the react site opens fine ... I click on login and it takes me to the zitadel.mysvr.local login/register page and allows me to start the auth flow

I am beginning to suspect that when using https, there is some failure to pass along the host header which zitadel needs to identify instances. Not sure where the disconnect is happening. I have tried everything I can think of. If one of your devs have a minimal (working) docker-compose + traefik sample I can do further troubleshooting with that. Alternatively clear documentation for how to get this working in a reverse proxied, local https setup would help as well. e.g docs have

>To allow ZITADEL to function properly please make sure that HTTP/2 is enabled. If you are using the mode external or disabled make sure to verify h2c compatibilty

How do we ensure that we have everything needed for reverse proxied https setup to work?

Much appreciated

**Comment by bernardro** [comment]
[this](https://github.com/authelia/authelia/blob/master/examples/compose/local/setup.sh) is exactly the kind of thing I am looking to setup 
```
...
Setup completed successfully.

You can now visit the following locations:
- https://public.$DOMAIN - Bypasses Authelia
- https://traefik.$DOMAIN - Secured with Authelia one-factor authentication
- https://secure.$DOMAIN - Secured with Authelia two-factor authentication (see note below)
...
```

**Comment by bernardro** [comment]
from what I read [here](https://stackoverflow.com/questions/46788904/why-do-web-browsers-not-support-h2c-http-2-without-tls) most browsers do not support h2c for [various reasons](https://bishopfox.com/blog/h2c-smuggling-request)?

**Comment by bernardro** [comment]
so I think my hunch was correct - when I add this to traefik.yaml
```
...
    forwardedHeaders:
      insecure: true
...
```
now, for the first time, I get a different error when I browse the react project (before setting it up in zitadel console)
> The requested redirect_uri is missing in the client configuration. If you have any questions, you may contact the administrator of the application.

the redirect_uri, however, does appear to be in the url
>https://zitadel.mysvr.local/oauth/v2/authorize?
response_type=code
&client_id=237671583051743238%40zitadel
&state=YXFQc0VjemkzemdlZEdJbkdNbWFkaEVnd2p2SE00MTI4MkdIQkw5bVRzazV2%3B76e297d9-1df4-4257-800f-eef7e9187536
&redirect_uri=https%3A%2F%2Fzitareact1.mysvr.local%2Fui%2Fconsole%2Fauth%2Fcallback
&scope=openid%20profile%20email
&code_challenge=PWFOxyjSk6w4N-420dTR9gWgB_Y0gnVnMbghthRGCxU
&code_challenge_method=S256
&nonce=YXFQc0VjemkzemdlZEdJbkdNbWFkaEVnd2p2SE00MTI4MkdIQkw5bVRzazV2

**Reply by bernardro** [reply]
I happen to be using
>ZITADEL_VERSION=v2.38.1

will switch to latest to see if anything changes

**Reply by hifabienne** [reply]
did you also configure the redirect url on your application? Those urls need to match

**Reply by bernardro** [reply]
yep

**Comment by bernardro** [comment]
when I visit `https://zitareact1.mysvr.local/.well-known/openid-configuration` at the bottom, I see
```
...
request_uri_parameter_supported: false
...
```

not sure why that is set to false
